crosscluster/ldrrandgen.TestGenerateLDRTable: TestGenerateLDRTable failed

crosscluster/ldrrandgen.TestGenerateLDRTable.TestGenerateLDRTable [failed](https://mesolite.cluster.engflow.com/invocations/default/403a188d-da2d-43be-9670-d7fbb37ab6a3?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY3Jvc3NjbHVzdGVyL2xkcnJhbmRnZW46bGRycmFuZGdlbl90ZXN0) on master @ [7f098a12bbaf8c31787db28b36f06c6e133c5c00](https://github.com/cockroachdb/cockroach/commits/7f098a12bbaf8c31787db28b36f06c6e133c5c00):


```
=== RUN   TestGenerateLDRTable
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestGenerateLDRTable1625358931
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:154: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    logical_test.go:47: creating table: CREATE TABLE test_writer (col0_0 BPCHAR NOT NULL, col0_1 NAME[] NOT NULL, col0_2 BIT, col0_3 UUID NOT NULL, col0_4 STRING NULL, col0_5 REFCURSOR NOT NULL, col0_6 BPCHAR NOT NULL, col0_7 INT8, col0_8 STRING NULL AS (lower(col0_4)) VIRTUAL, col0_9 INT8 AS (abs(col0_7)) STORED, PRIMARY KEY (col0_0 DESC), INDEX (col0_6, col0_9), INDEX (col0_3 DESC))
    logical_test.go:57: error executing query="CREATE LOGICALLY REPLICATED TABLE b.test_writer FROM TABLE a.test_writer ON 'external://a' WITH BIDIRECTIONAL ON 'external://b'" args=[]: pq: cannot create logical replication stream: column col0_5 is a RefCursor
    panic.go:615: -- test log scope end --
test logs left over in: outputs.zip/logTestGenerateLDRTable1625358931
--- FAIL: TestGenerateLDRTable (1.59s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestGenerateLDRTable.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-56042

crosscluster/ldrrandgen.TestGenerateLDRTable.TestGenerateLDRTable [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20672066?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20672066?buildTab=artifacts#/) on master @ [f0f7fd8ac9fdd78bd6c6b7f0c80cf2caf9bd2665](https://github.com/cockroachdb/cockroach/commits/f0f7fd8ac9fdd78bd6c6b7f0c80cf2caf9bd2665):


```
=== RUN   TestGenerateLDRTable
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/1ee0ce66df5f0964c8d393c1246a3667/logTestGenerateLDRTable4151198464
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:154: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    logical_test.go:47: creating table: CREATE TABLE test_writer (col0_0 INET NOT NULL, col0_1 TSVECTOR, col0_2 UUID NOT NULL, col0_3 INET NOT NULL, col0_4 DATE NOT NULL, col0_5 REFCURSOR NOT NULL, col0_6 NAME NOT NULL, col0_7 VECTOR NOT NULL, col0_8 BIT(30), col0_9 REFCURSOR NOT NULL, col0_10 NAME, col0_11 STRING NOT NULL AS (lower(CAST(col0_7 AS STRING))) STORED, col0_12 STRING AS (lower(CAST(col0_8 AS STRING))) STORED, col0_13 STRING AS (lower(CAST(col0_1 AS STRING))) VIRTUAL, col0_14 STRING NOT NULL AS (CASE WHEN col0_4 IS NULL THEN e'\x1b.\x06\x05\x011S;' ELSE 'c&6|a2j' END) VIRTUAL, col0_15 STRING AS (lower(CAST(col0_8 AS STRING))) STORED, PRIMARY KEY (col0_4 DESC), INVERTED INDEX (col0_6 gin_trgm_ops))
    logical_test.go:57: error executing query="CREATE LOGICALLY REPLICATED TABLE b.test_writer FROM TABLE a.test_writer ON 'external://a' WITH BIDIRECTIONAL ON 'external://b'" args=[]: pq: cannot create logical replication stream: column col0_5 is a RefCursor
    panic.go:615: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/1ee0ce66df5f0964c8d393c1246a3667/logTestGenerateLDRTable4151198464
--- FAIL: TestGenerateLDRTable (1.10s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestGenerateLDRTable.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/crosscluster/ldrrandgen/logical.go
+++ b/pkg/crosscluster/ldrrandgen/logical.go
@@ -35,6 +35,11 @@ func GenerateLDRTable(
 			if columnDef.Type.(*types.T).Family() == types.ArrayFamily && columnDef.Type.(*types.T).ArrayContents().Family() == types.BitFamily {
 				return false
 			}
+			// We don't support RefCursor columns in LDR tables because they do not
+			// support equality.
+			if columnDef.Type.(*types.T).Family() == types.RefCursorFamily {
+				return false
+			}
 			// We don't allow the special '"char"' column because pgwire truncates the value to 1 byte.
 			// TODO(jeffswenson): remove this once #149427 is fixed.
 			if columnDef.Type.(*types.T).Family() == types.StringFamily && columnDef.Type.(*types.T).Width() == 1 {

